File: day05/01_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func match(s1, s2 string) bool {
	return strings.ToLower(s1) == strings.ToLower(s2) && s1 != s2
}

func produce(str string, ch chan string) {
	l := len(str)
	if l == 0 {
		ch <- ""
	} else if l == 1 {
		ch <- str
	} else {
		half := int(l / 2)

		leftChan := make(chan string)
		rightChan := make(chan string)

		go produce(str[:half], leftChan)
		go produce(str[half:], rightChan)

		ch <- merge(
			<-leftChan,
			<-rightChan,
		)
	}
}

func merge(front, back string) string {
	if front == "" && back == "" {
		return ""
	} else if front == "" {
		return back
	} else if back == "" {
		return front
	} else if match(front[len(front)-1:], back[0:1]) {
		return merge(front[:len(front)-1], back[1:])
	} else {
		return front + back
	}
}

func main() {
	file, _ := os.Open("input.txt")
	// file, _ := os.Open("sample.txt")

	defer file.Close()

	scanner := bufio.NewScanner(file)

	var str string

	for scanner.Scan() {
		str = scanner.Text()
	}

	ch := make(chan string)

	go produce(str, ch)
	ans := <-ch
	fmt.Println(len(ans))
}

File: day05/01_1.js
const fs = require('fs')

function match(s1, s2) {
  return s1.toLowerCase() == s2.toLowerCase() && s1 !== s2
}

function produce(str) {
  const l = str.length

  if (l === 0) {
    return ''
  } else if (l === 1) {
    return str
  } else {
    const half = Math.round(l / 2)

    const left = str.slice(0, half)
    const right = str.slice(half)

    return merge(produce(left), produce(right))
  }
}

function merge(left, right) {
  if (left === '' && right === '') {
    return ''
  } else if (left === '') {
    return right
  } else if (right === '') {
    return left
  } else if (match(left.charAt(left.length - 1), right.charAt(0))) {
    return merge(left.slice(0, left.length - 1), right.slice(1))
  } else {
    return left + right
  }
}

// const data = fs.readFileSync('sample.txt', 'utf8')
const data = fs.readFileSync('input.txt', 'utf8')

const ans = produce(data).length
console.log(ans)